SQL: Fix literal projection with condition

Luegg opened the pull request:
Fixes https://github.com/elastic/elasticsearch/issues/64567

Queries with a literal selection and a filter like `SELECT 1 FROM test_emp WHERE gender = 'F'` are currently erroneously optimised to use a local relation. This causes ES to always return a single record, no matter how many records match the filter condition.

This PR makes sure that `SkipQueryIfFoldingProjection` only skips the query if it's an aggregate with only constants (e.g. `SELECT 'foo' FROM test GROUP BY 1`). This optimization seems to lead to another issue https://github.com/elastic/elasticsearch/issues/74064 that's not yet addressed in this PR.

Besides this the "skip query" optimization, the `SkipQueryIfFoldingProjection` class also folds constants from `LocalRelation`s and pushes the evaluated values into a new `LocalRelation` (e.g. for queries like `SELECT 1 + 2`).

Commit:
fix literal projections of es index with filter

Commit:
sql spec

Commit:
Merge master into fix/literalProjectionWithCondition

Commit:
fix imports

Commit:
fix all imports

Commit:
replace asserts

Commit:
fmt

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1149,22 +1150,48 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
-                List<Object> values = extractConstants(p.projections());
-                if (values.size() == p.projections().size() && (p.child() instanceof EsRelation) == false &&
-                    isNotQueryWithFromClauseAndFilterFoldedToFalse(p)) {
+            var leafRelation = leafRelation(plan);
```
Personally, I'm not fun of `var` in the code, I like more in the tests for simplification, but let's wait for the opinion of the others.

bpintea replied:
Personally I also find that making the type explicit is more readable, although there are places in ES - less so in *QL, I think - where `var` is used.

Luegg replied:
👍 Costin also pointed out that `var` is not yet available in Java 8, so not using it also helps a lot with backporting the fix to 7.x.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1149,22 +1150,48 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
-                if (values.size() == a.aggregates().size() && a.groupings().isEmpty()
-                    && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
+
+                // aggregations on only constant values like "SELECT 'foo' FROM test GROUP BY 1"
```
Might be worth adding this statement (or similar) into the optimiser tests too, imo.

Luegg replied:
`testSkipQueryOnEsAggregationWithOnlyConstants` should cover this case

bpintea reviewed (commented):
Nice! Looks good, only left a few small notes.

Commit:
use explicit types instead of vars

Commit:
more specs

Commit:
use collectFirstDown to bail out early

Commit:
avoid null

Commit:
no List.of

Commit:
clearer spec name

Commit:
rm unused import

Luegg commented:
I've addressed your comments and also tried to consolidate the partially duplicated logic of the `Project` and `Aggregate` cases. Thanks to the consolidated logic and the `collectFirstDown` method it should also use less traversals through the query tree.

Please have another look.

Commit:
fix non-relation leaf case

Commit:
Merge origin/master into fix/literalProjectionWithCondition

matriv reviewed (approved):
LGTM. Thank you!

astefan commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/tree/Node.java:
```diff
@@ -161,6 +162,33 @@ protected void doCollectFirst(Predicate<? super T> predicate, List<T> matches) {
         }
     }
 
+    public <R> Optional<R> collectFirstDown(Class<R> typeToken) {
```
This method is never used.
I'd suggest adding it when needed.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1148,32 +1149,37 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
+                    Aggregate a = (Aggregate) p;
+                    List<Object> folded = extractConstants(a.aggregates());
+
+                    boolean onlyConstantAggregations = relation instanceof EsRelation &&
```
We should try to use a consistent code style and one of the practices is the use of `&&` and `||` at the start of the line, not at the end, in case of multiline conditions.

astefan reviewed (approved):
LGTM overall. Left two minor comments.
Also, why the use of `Optional`? It's not something we use in *QL.

bpintea reviewed (approved):
Lgtm.

Commit:
address comments

Commit:
Merge master into fix/literalProjectionWithCondition

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/tree/Node.java:
```diff
@@ -161,6 +162,23 @@ protected void doCollectFirst(Predicate<? super T> predicate, List<T> matches) {
     }
 
+    @SuppressWarnings("unchecked")
+    public <R> Optional<R> collectFirstDown(Function<T, Optional<R>> fn) {
```
This method is not needed; it's a tweaked version of `collectFirstChildren` that has several issues:

1. collect methods should return a list not a single value.
2. Optional is not used in the QL code-base (missing is null for single value, empty lists for multiple ones).
3. for condition parameter, use Predicate not Function (another reason why Optional doesn't work)
4. collectFirstDown is similar to collectFirstChildren. The former returns at first match while the latter looks at all current children in case of a match. I don't think there are any performance issues with this approach and it forces the caller to look at the issue in a more general fashion:
- do I only need for one child to match?
- what if there are multiple matches - is that acceptable or a non existing case?

costin commented on x-pack/plugin/sql/qa/server/src/main/resources/select.sql-spec:
```diff
@@ -54,6 +54,10 @@ sameConstantsWithLimitV2
+constantWithWhere
+SELECT 1 FROM "test_emp" WHERE gender = 'F';
+constantWithWhereOrderAndLimit
+SELECT 1 FROM "test_emp" WHERE gender = 'F' ORDER BY emp_no LIMIT 5;
```
Please add more tests to test different varieties of tested plan:
`WHERE a and b`
`WHERE a and function(b)`
`WHERE a and b ORDER BY c`

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1148,32 +1149,37 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
-                    optimizedPlan.set(new LocalRelation(p.source(), new SingletonExecutable(p.output(), values.toArray())));
-                }
-            });
+            LogicalPlan relation = plan
```
This statement is confusing since it looks for nodes that have no children. Which currently happens to be `Relation` nodes however that's not mandatory in the future.
It's better to just introduce a `Relation` class which can be just an extension to `LeafPlan` and be the super class of all existing Relation classes (`EsRelation` and `LocalRelation`).
Then this condition becomes `plan.collect(Relation::instanceOf)`.
As an alternative that doesn't require introducing a `Relation` class simply look at the leaves and check whether they are relationship - `plan.collectLeaves()`.

costin reviewed (changes requested):
Left a few comments regarding the new method on `Node` and iteration style.

Commit:
address comments

Commit:
collect leaves instead of relations

Luegg commented:
@costin please have another look, I've addressed your comments and the implementation should now also be more robust against some cases where the optimization might have erroneously be applied when it shouldn't. E.g. there might be cases (in the future?) with more than one Project/Aggregate or Relation.

I'm hesitant of introducing `Relation` as I'm not sure it adds much value in general and even in this case I'm only interested in two specific relations and not e.g. `UnresolvedRelation` or the EQL `LocalRelation`. The `collectLeaves` approach seems sufficient to me.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1148,29 +1148,40 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
-                if (values.size() == a.aggregates().size() && a.groupings().isEmpty()
-                    && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
-                    optimizedPlan.set(new LocalRelation(a.source(), new SingletonExecutable(a.output(), values.toArray())));
+                if (projectOrAggregate instanceof Project && isNonSkippedLocalRelation) {
```
The if/else could be optimized by swapping it around and avoid checking for Project:
```
if (projectOrAggregate instanceof Aggregate) {
 ...
 }
else if (isNonSkippedLocalRelation) // no need to check for Project
```

costin reviewed (approved):
Thanks. This looks great!
